Sources/SwiftProtobuf/TextFormatEncodingVisitor.swift

```diff
@@ -14,209 +14,409 @@
+  }
+
+  mutating func visitPackedFixed64Field(value: [UInt64], fieldNumber: Int) throws {
+    try visitPackedUInt64Field(value: value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitPackedSFixed32Field(value: [Int32], fieldNumber: Int) throws {
+    try visitPackedInt32Field(value: value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitPackedSFixed64Field(value: [Int64], fieldNumber: Int) throws {
+    try visitPackedInt64Field(value: value, fieldNumber: fieldNumber)
+  }
+
+  mutating func visitPackedBoolField(value: [Bool], fieldNumber: Int) throws {
+    try _visitPacked(value: value, fieldNumber: fieldNumber) { (v: Bool) in
+      encoder.putBoolValue(value: v)
+    }
+  }
+
+  mutating func visitPackedEnumField<E: Enum>(value: [E], fieldNumber: Int) throws {
+    try _visitPacked(value: value, fieldNumber: fieldNumber) { (v: E) in
+      encoder.putEnumValue(value: v)
+    }
+  }
+
+  /// Helper to encapsulate the common structure of iterating over a map
+  /// and encoding the keys and values.
+  private mutating func _visitMap<K, V>(
+    map: Dictionary<K, V>,
+    fieldNumber: Int,
+    coder: (inout TextFormatEncodingVisitor, K, V) throws -> ()
+  ) throws {
     let protoFieldName = try self.protoFieldName(for: fieldNumber)
-    for v in value {
-      encoder.startMessageField(name: protoFieldName)
+    for (k,v) in map {
+      encoder.startMessageField(name: protoFieldName, inExtension: inExtension)
       encoder.startObject()
-      let visitor = TextFormatEncodingVisitor(message: v, encoder: encoder)
-      try v.traverse(visitor: visitor)
+      var visitor = TextFormatEncodingVisitor(nameResolver: mapNameResolver, encoder: encoder)
+      try coder(&visitor, k, v)
+      encoder = visitor.encoder
       encoder.endObject()
       encoder.endField()
     }
   }
 
-  func visitMapField<KeyType: MapKeyType, ValueType: MapValueType>(
+  mutating func visitMapField<KeyType: MapKeyType, ValueType: MapValueType>(
     fieldType: ProtobufMap<KeyType, ValueType>.Type,
     value: ProtobufMap<KeyType, ValueType>.BaseType,
     fieldNumber: Int
   ) throws where KeyType.BaseType: Hashable {
-    let protoFieldName = try self.protoFieldName(for: fieldNumber)
-    for (k,v) in value {
-      encoder.startMessageField(name: protoFieldName)
-      encoder.startObject()
-      encoder.startField(name: "key")
-      try KeyType.serializeTextValue(encoder: encoder, value: k)
-      encoder.endField()
-      encoder.startField(name: "value")
-      try ValueType.serializeTextValue(encoder: encoder, value: v)
-      encoder.endField()
-      encoder.endObject()
-      encoder.endField()
-    }
+      try _visitMap(map: value, fieldNumber: fieldNumber) {
+          (visitor: inout TextFormatEncodingVisitor, key, value) throws -> () in
+          try KeyType.visitSingular(value: key, fieldNumber: 1, with: &visitor)
+          try ValueType.visitSingular(value: value, fieldNumber: 2, with: &visitor)
```

Comment:
Note the maps here are using the general-purpose reflectors on FieldType, which is what allows us to lose the specialized `serializeTextValue` methods we used to have.